### ISSUE TITLE
8B+ eval risk scores not bitwise-reproducible across GPU nodes (~1e-5 long-sequence float non-determinism)

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/cruijff_kit#547]
## Summary

8B+ eval **risk scores are not bitwise-reproducible across GPU nodes**. The same model, same data, same decoding config produces risk scores that differ at the ~1e-5 level when the eval runs on a different physical node. Completions (the generated text) are always identical; only the float-valued logprob-derived risk scores and the calibration metrics aggregated from them drift.

This surfaced while verifying #499 (`do_sample=false` default). It is **not** caused by that change — see the control below — and predates it. Filing so it's recorded rather than rediscovered.

## Evidence

ACS-income, Llama-3.1-8B-Instruct, 5,000 samples, `do_sample=false` held **constant** across all runs, varying only the node:

| Comparison | Completions | Risk scores | Aggregate metrics |
|------------|-------------|-------------|-------------------|
| **Cross-node** (della-l08g2 vs della-l09g3) | 5000/5000 identical | 3380/5000 identical — 1620 diffs, max \|Δ\| 2.73e-5, mean 5.14e-6 | 5/5 differ |
| **Same-node** (della-l09g3 vs della-l09g3) | 5000/5000 identical | **5000/5000 — bitwise identical** | 0/5 differ |

Same node → perfectly deterministic. Different node → ~1e-5 wobble. The decoding mode is identical in both rows, so this isolates the cause to **hardware/reduction-order float non-determinism**, not sampling.

## What we believe is happening

- Different GPU nodes (even nominally identical SKUs) can use different reduction orders / kernel selections for the matmuls feeding the final logits.
- Floating-point addition isn't associative, so long sequences accumulate tiny per-token differences.
- `risk_scorer` reads logprobs over the option tokens, so those differences land directly in `risk_score` and propagate into `mean_risk_score`, `brier_score`, `ECE`, `risk_calibration_error`, `auc_score`.

## Scope / severity

- **Magnitude:** ~1e-5 per sample. Almost certainly below any scientific decision threshold for our experiments.
- **Length-dependent:** short-sequence tasks (e.g. `bits_majority`, ~15 tokens) are immune even cross-node. This only bites long-context inference.
- **Model-size-dependent observed so far:** seen on 8B; 1B/3B ACS comparisons were identical in the #499 verification (though those weren't node-pinned controls).
- **Completions unaffected:** argmax decoding is robust to the perturbation; only continuous logprob readouts move.

## Possible directions (if we decide it's worth it)

1. **Document & accept** — note in `KNOWN_ISSUES.md` that cross-node risk scores are reproducible only to ~1e-5; pin nodes when bitwise reproducibility is required.
2. **Pin nodes for reproducibility-critical runs** — add an optional `--nodelist` knob to eval scaffolding.
3. **Investigate determinism flags** — `torch.use_deterministic_algorithms`, `CUBLAS_WORKSPACE_CONFIG`, etc., and measure whether they close the gap (they may not, cross-hardware).
4. **Do nothing** — if 1e-5 is comfortably below our decision resolution, this may be a "known and fine" footnote.

## Reproduction artifacts

Node-pinned eval logs + comparator at `tmp/dosample_acs_nondet_2026-06-02/` (compare.py prints the table above).

---

@msalganik @sarahepedersen — is this worth investigating, or do we log it as a known ~1e-5 footnote and move on? The magnitude looks scientifically negligible to me, but I wanted it on the record before the details faded, and you two may have a sharper sense of whether downstream analyses ever lean on risk-score reproducibility at that resolution.

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Closing as part of the September 2026 archive triage.

No one weighed in, so the verdict defaults to option 1, document and accept: a ~1e-5 cross-node wobble in logprob-derived scores is well below any decision threshold these experiments used, and completions are unaffected. Recorded in the handoff notes as a known property of long-sequence 8B evals, with the node-pinning workaround. The determinism flags in option 3 were never tested.

—MxC
